lib/fizzy/execute.cpp

```diff
@@ -187,42 +190,41 @@ std::tuple<bytes_ptr, size_t> allocate_memory(const std::vector<Memory>& module_
 
     if (module_memories.size() == 1)
     {
-        const size_t memory_min = module_memories[0].limits.min;
-        const size_t memory_max =
-            (module_memories[0].limits.max.has_value() ? *module_memories[0].limits.max :
-                                                         MemoryPagesLimit);
+        const auto memory_min = module_memories[0].limits.min;
+        const auto memory_max = module_memories[0].limits.max;
+
         // FIXME: better error handling
-        if ((memory_min > MemoryPagesLimit) || (memory_max > MemoryPagesLimit))
+        if ((memory_min > MemoryPagesLimit) ||
+            (memory_max.has_value() && *memory_max > MemoryPagesLimit))
         {
             throw instantiate_error("Cannot exceed hard memory limit of " +
                                     std::to_string(MemoryPagesLimit * PageSize) + " bytes.");
         }
 
         // NOTE: fill it with zeroes
         bytes_ptr memory{new bytes(memory_min * PageSize, 0), bytes_delete};
-        return {std::move(memory), memory_max};
+        return {std::move(memory), module_memories[0].limits};
     }
     else if (imported_memories.size() == 1)
     {
-        const size_t memory_min = imported_memories[0].limits.min;
-        const size_t memory_max =
-            (imported_memories[0].limits.max.has_value() ? *imported_memories[0].limits.max :
-                                                           MemoryPagesLimit);
+        const auto memory_min = imported_memories[0].limits.min;
+        const auto memory_max = imported_memories[0].limits.max;
 
         // FIXME: better error handling
-        if ((memory_min > MemoryPagesLimit) || (memory_max > MemoryPagesLimit))
+        if ((memory_min > MemoryPagesLimit) ||
+            (memory_max.has_value() && *memory_max > MemoryPagesLimit))
         {
             throw instantiate_error("Imported memory limits cannot exceed hard memory limit of " +
                                     std::to_string(MemoryPagesLimit * PageSize) + " bytes.");
         }
 
         bytes_ptr memory{imported_memories[0].data, null_delete};
-        return {std::move(memory), memory_max};
+        return {std::move(memory), imported_memories[0].limits};
     }
     else
     {
         bytes_ptr memory{nullptr, null_delete};
-        return {std::move(memory), MemoryPagesLimit};
+        return {std::move(memory), Limits{}};
```

Comment:
No need to set any limit when memory is not present, because `memory.grow` would be invalid.